cloudinit/sources/DataSourceAzure.py

```diff
@@ -982,7 +985,30 @@ def _report_ready_for_pps(self) -> None:
             report_diagnostic_event(msg, logger_func=LOG.error)
             raise sources.InvalidMetaDataException(msg) from error
 
-        self._create_report_ready_marker()
+        if create_marker:
+            self._create_report_ready_marker()
+
+    @azure_ds_telemetry_reporter
+    def _shutdown_for_os_disk_pps(self):
+        """Shutdown the VM gracefully or wait until host forces it."""
+        self._report_ready_for_pps(create_marker=False)
+        shutdown_cmd = self.distro.shutdown_command(
+            mode="poweroff", delay="now", message="PPS OS disk"
+        )
+        report_diagnostic_event(
+            "Shutting down pre-provisioned VM with cmd=%r" % shutdown_cmd,
+            logger_func=LOG.info,
+        )
+        try:
+            subp.subp(shutdown_cmd)
+        except subp.ProcessExecutionError:
+            report_diagnostic_event(
+                "Shutdown command failed, waiting for host to force it...",
+                logger_func=LOG.info,
+            )
+
+        sleep(31536000)
+        raise BrokenAzureDataSource("Shutdown failure for PPS disk.")
```

Comment:
does this exception lead to report failure? If possible, reporting failure would be best if it cannot be shutdown

Reply:
nvm, can't report failure after reporting ready

Reply:
It would go down that code path anyways.